// @lrnwebcomponents/app-hax/lib/v1/app-hax-steps.js
define(["exports", "../../../../lit/index.js", "./AppHaxRouter.js", "../../../simple-colors/simple-colors.js", "../../../../mobx/dist/mobx.esm.js", "./AppHaxStore.js", "../../../utils/utils.js", "../../../../scrollable-component/index.js", "./app-hax-site-button.js", "./app-hax-hat-progress.js", "./app-hax-button.js"], function (_exports, _index, _AppHaxRouter, _simpleColors, _mobxEsm, _AppHaxStore, _utils, _index2, _appHaxSiteButton, _appHaxHatProgress, _appHaxButton) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.AppHaxSteps = void 0;

  var _templateObject_d77b49e0c0ce11ecbdb54bb05e9423ad, _templateObject2_d77b49e0c0ce11ecbdb54bb05e9423ad, _templateObject3_d77b49e0c0ce11ecbdb54bb05e9423ad, _templateObject4_d77b49e0c0ce11ecbdb54bb05e9423ad, _templateObject5_d77b49e0c0ce11ecbdb54bb05e9423ad, _templateObject6_d77b49e0c0ce11ecbdb54bb05e9423ad, _templateObject7_d77b49e0c0ce11ecbdb54bb05e9423ad;

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  var AppHaxSteps = /*#__PURE__*/function (_SimpleColors) {
    babelHelpers.inherits(AppHaxSteps, _SimpleColors);

    var _super = _createSuper(AppHaxSteps);

    function AppHaxSteps() {
      var _this;

      babelHelpers.classCallCheck(this, AppHaxSteps);
      _this = _super.call(this);
      _this.stepRoutes = [];
      _this._progressReady = false;
      _this.step = null;
      _this.loaded = false;
      _this.themeNames = [];
      _this.appSettings = {};
      (0, _mobxEsm.autorun)(function () {
        _this.appSettings = (0, _mobxEsm.toJS)(_AppHaxStore.store.appSettings);
        _this.themeNames = Object.keys(_this.appSettings.themes);
      });
      (0, _mobxEsm.autorun)(function () {
        (0, _utils.localStorageSet)('app-hax-step', (0, _mobxEsm.toJS)(_AppHaxStore.store.step));
      });
      (0, _mobxEsm.autorun)(function () {
        (0, _utils.localStorageSet)('app-hax-site', (0, _mobxEsm.toJS)(_AppHaxStore.store.site));
        _this.step = _AppHaxStore.store.stepTest(_this.step);
      });
      (0, _mobxEsm.autorun)(function () {
        if ((0, _mobxEsm.toJS)(_AppHaxStore.store.createSiteSteps)) {
          var location = (0, _mobxEsm.toJS)(_AppHaxStore.store.location);
          _this.step = _AppHaxStore.store.stepTest(_this.step);
        }
      }); // routes, but only the ones that have a step property

      (0, _mobxEsm.autorun)(function () {
        var routes = (0, _mobxEsm.toJS)(_AppHaxStore.store.routes);
        _this.stepRoutes = routes.filter(function (item) {
          return item.step;
        });
      });
      return _this;
    }

    babelHelpers.createClass(AppHaxSteps, [{
      key: "chooseStructure",
      value: // step 1
      function chooseStructure(e) {
        var value = e.target.value;
        _AppHaxStore.store.site.structure = value;

        _AppHaxStore.store.appEl.playSound('click2');
      } // step 2

    }, {
      key: "chooseType",
      value: function chooseType(e) {
        var value = e.target.value;
        _AppHaxStore.store.site.type = value;

        _AppHaxStore.store.appEl.playSound('click2');
      } // step 3

    }, {
      key: "chooseTheme",
      value: function chooseTheme(e) {
        var value = e.target.value;
        _AppHaxStore.store.site.theme = value;

        _AppHaxStore.store.appEl.playSound('click2');
      } // step 4

    }, {
      key: "chooseName",
      value: function chooseName(e) {
        var value = this.shadowRoot.querySelector("#sitename").value;
        _AppHaxStore.store.site.name = value;

        _AppHaxStore.store.appEl.playSound('click2');
      }
    }, {
      key: "progressReady",
      value: function progressReady(e) {
        var _this2 = this;

        if (e.detail) {
          this._progressReady = true;

          if (this.step === 5) {
            setTimeout(function () {
              _this2.shadowRoot.querySelector('app-hax-hat-progress').process();
            }, 300);
          }
        }
      }
    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this3 = this;

        if (babelHelpers.get(babelHelpers.getPrototypeOf(AppHaxSteps.prototype), "updated", this)) {
          babelHelpers.get(babelHelpers.getPrototypeOf(AppHaxSteps.prototype), "updated", this).call(this, changedProperties);
        }

        changedProperties.forEach(function (oldValue, propName) {
          // set input field to whats in store if we have it
          if (_this3.step === 4 && propName === 'step' && _this3.shadowRoot) {
            _this3.shadowRoot.querySelector('#sitename').value = (0, _mobxEsm.toJS)(_AppHaxStore.store.site.name);
          } // progress


          if (_this3.step === 5 && propName === 'step' && _this3.shadowRoot && _this3._progressReady) {
            setTimeout(function () {
              _this3.shadowRoot.querySelector('app-hax-hat-progress').process();
            }, 600);
          } // update the store for step when it changes internal to our step flow


          if (propName === 'step') {
            _AppHaxStore.store[propName] = _this3[propName];
          }
        });
      }
    }, {
      key: "connectedCallback",
      value: function connectedCallback() {
        babelHelpers.get(babelHelpers.getPrototypeOf(AppHaxSteps.prototype), "connectedCallback", this).call(this);
        window.addEventListener('resize', this.maintainScroll.bind(this));
      }
    }, {
      key: "disconnectedCallback",
      value: function disconnectedCallback() {
        window.removeEventListener('resize', this.maintainScroll.bind(this));
        babelHelpers.get(babelHelpers.getPrototypeOf(AppHaxSteps.prototype), "disconnectedCallback", this).call(this);
      } // account for resizing

    }, {
      key: "maintainScroll",
      value: function maintainScroll() {
        var _this4 = this;

        if (this.shadowRoot && this.step) {
          this.scrollToThing("#step-".concat(this.step), {
            behavior: 'instant',
            block: 'nearest',
            inline: 'nearest'
          }); // account for an animated window drag... stupid.

          setTimeout(function () {
            _this4.scrollToThing("#step-".concat(_this4.step), {
              behavior: 'instant',
              block: 'nearest',
              inline: 'nearest'
            });
          }, 100);
        }
      }
    }, {
      key: "firstUpdated",
      value: function firstUpdated(changedProperties) {
        var _this5 = this;

        if (babelHelpers.get(babelHelpers.getPrototypeOf(AppHaxSteps.prototype), "firstUpdated", this)) {
          babelHelpers.get(babelHelpers.getPrototypeOf(AppHaxSteps.prototype), "firstUpdated", this).call(this, changedProperties);
        }

        setTimeout(function () {
          // ensure paint issues not a factor for null step
          if (_this5.step === null) {
            _this5.step = 1;
          }

          _this5.scrollToThing("#step-".concat(_this5.step), {
            behavior: 'instant',
            block: 'nearest',
            inline: 'nearest'
          });
        }, 100);
        (0, _mobxEsm.autorun)(function () {
          // verify we are in the site creation process
          if ((0, _mobxEsm.toJS)(_AppHaxStore.store.createSiteSteps) && (0, _mobxEsm.toJS)(_AppHaxStore.store.appReady)) {
            var location = (0, _mobxEsm.toJS)(_AppHaxStore.store.location);

            if (location.route && location.route.step && location.route.id) {
              // account for an animated window drag... stupid.
              setTimeout(function () {
                _this5.scrollToThing('#'.concat(location.route.id), {
                  block: 'nearest',
                  inline: 'nearest',
                  behavior: 'smooth'
                });
              }, 300); // this delay helps w/ initial paint timing but also user perception
              // there's a desire to have a delay especialy when tapping things of
              // about 300ms
            }
          }
        });
        (0, _mobxEsm.autorun)(function () {
          if (_this5.shadowRoot && (0, _mobxEsm.toJS)(_AppHaxStore.store.createSiteSteps) && (0, _mobxEsm.toJS)(_AppHaxStore.store.appReady)) {
            var activeItem = (0, _mobxEsm.toJS)(_AppHaxStore.store.activeItem);

            if (activeItem && activeItem.step && !_this5.__overrideProgression) {
              _this5.shadowRoot.querySelector('#link-'.concat(activeItem.id)).click();
            }
          }
        });
      }
      /**
       * Yet another reason Apple doesn't let us have nice things.
       * This detects the NONSTANDARD BS VERSION OF SCROLLINTOVIEW
       * and then ensures that it incorrectly calls to scroll into view
       * WITHOUT the wonderful params that ALL OTHER BROWSERS ACCEPT
       * AND MAKE OUR LIVES SO WONDERFUL TO SCROLL TO THINGS SMOOTHLY
       */

    }, {
      key: "scrollToThing",
      value: function scrollToThing(sel, props) {
        var isSafari = window.safari !== undefined;

        if (this.shadowRoot.querySelector('.carousel-with-snapping-item.active-step')) {
          this.shadowRoot.querySelector('.carousel-with-snapping-item.active-step').classList.remove('active-step');
        }

        if (isSafari) {
          this.shadowRoot.querySelector(sel).scrollIntoView();
        } else {
          this.shadowRoot.querySelector(sel).scrollIntoView(props);
        }

        this.shadowRoot.querySelector(sel).classList.add('active-step');
      }
    }, {
      key: "progressFinished",
      value: function progressFinished(e) {
        if (e.detail) {
          this.loaded = true;

          _AppHaxStore.store.appEl.playSound('success');
        }
      }
    }, {
      key: "stepLinkClick",
      value: function stepLinkClick(e) {
        var clickedStep = parseInt(e.target.getAttribute("data-step"));

        if (this.step < clickedStep) {
          e.preventDefault();
        } // means user went backwards
        else if (this.step > clickedStep) {
            if (clickedStep === 1) {
              _AppHaxStore.store.site.structure = null;
              _AppHaxStore.store.site.type = null;
              _AppHaxStore.store.site.theme = null;
            } else if (clickedStep === 2) {
              _AppHaxStore.store.site.type = null;
              _AppHaxStore.store.site.theme = null;
            } else if (clickedStep === 3) {
              _AppHaxStore.store.site.theme = null;
            }

            this.step = clickedStep;
          }
      }
    }, {
      key: "renderTypes",
      value: function renderTypes(step) {
        var structure = (0, _mobxEsm.toJS)(_AppHaxStore.store.site.structure);
        var template = (0, _index.html)(_templateObject_d77b49e0c0ce11ecbdb54bb05e9423ad || (_templateObject_d77b49e0c0ce11ecbdb54bb05e9423ad = babelHelpers.taggedTemplateLiteral([""])));

        switch (structure) {
          case 'portfolio':
            template = (0, _index.html)(_templateObject2_d77b49e0c0ce11ecbdb54bb05e9423ad || (_templateObject2_d77b49e0c0ce11ecbdb54bb05e9423ad = babelHelpers.taggedTemplateLiteral(["\n        <app-hax-button\n          tabindex=\"", "\"\n          @click=", "\n          type=\"Technology\"\n        ></app-hax-button>\n        <app-hax-button\n          tabindex=\"", "\"\n          @click=", "\n          type=\"Business\"\n        ></app-hax-button>\n        <app-hax-button\n          tabindex=\"", "\"\n          @click=", "\n          type=\"Art\"\n        ></app-hax-button>\n        <app-hax-button\n          tabindex=\"", "\"\n          @click=", "\n        ></app-hax-button>"])), step !== 2 ? '-1' : '', this.chooseType, step !== 2 ? '-1' : '', this.chooseType, step !== 2 ? '-1' : '', this.chooseType, step !== 2 ? '-1' : '', this.chooseType);
            break;

          default:
          case 'course':
            template = (0, _index.html)(_templateObject3_d77b49e0c0ce11ecbdb54bb05e9423ad || (_templateObject3_d77b49e0c0ce11ecbdb54bb05e9423ad = babelHelpers.taggedTemplateLiteral(["\n        <app-hax-button\n          tabindex=\"", "\"\n          @click=", "\n          type=\"6w\"\n        ></app-hax-button>\n        <app-hax-button\n          tabindex=\"", "\"\n          @click=", "\n          type=\"15w\"\n        ></app-hax-button>\n        <app-hax-button\n          tabindex=\"", "\"\n          @click=", "\n          type=\"training\"\n        ></app-hax-button>\n        <app-hax-button\n          tabindex=\"", "\"\n          @click=", "\n        ></app-hax-button>"])), step !== 2 ? '-1' : '', this.chooseType, step !== 2 ? '-1' : '', this.chooseType, step !== 2 ? '-1' : '', this.chooseType, step !== 2 ? '-1' : '', this.chooseType);
            break;
        }

        return template;
      }
    }, {
      key: "render",
      value: function render() {
        var _this6 = this;

        return (0, _index.html)(_templateObject4_d77b49e0c0ce11ecbdb54bb05e9423ad || (_templateObject4_d77b49e0c0ce11ecbdb54bb05e9423ad = babelHelpers.taggedTemplateLiteral(["\n      <div id=\"container\">\n        <ul id=\"step-links\">\n          ", "\n        </ul>\n        <scrollable-component>\n          <div class=\"carousel-with-snapping-track\">\n            <div class=\"carousel-with-snapping-item\" id=\"step-1\">\n              <div class=\"step-wrapper\">\n                <app-hax-site-button\n                  tabindex=\"", "\"\n                  label=\"> Course\"\n                  value=\"course\"\n                  @click=", "\n                ></app-hax-site-button>\n                <app-hax-site-button\n                  tabindex=\"", "\"\n                  label=\"> Portfolio\"\n                  value=\"portfolio\"\n                  @click=", "\n                ></app-hax-site-button>\n              </div>\n            </div>\n            <div class=\"carousel-with-snapping-item\" id=\"step-2\">\n              <div id=\"grid-container\">\n                ", "\n              </div>\n          </div>\n          <div class=\"carousel-with-snapping-item\" id=\"step-3\">\n              <div id=\"themeContainer\">\n                ", "\n              </div>\n            </div>\n            <div class=\"carousel-with-snapping-item\" id=\"step-4\">\n              <input id=\"sitename\"\n              tabindex=\"", "\">\n              <app-hax-site-button\n                  class=\"sitenamebtn\"\n                  tabindex=\"", "\"\n                  label=\"> Create journey\"\n                  @click=", "\n                ></app-hax-site-button>\n            </div>\n            <div class=\"carousel-with-snapping-item\" id=\"step-5\">\n              <app-hax-hat-progress\n                @progress-ready=\"", "\"\n                @promise-progress-finished=\"", "\"\n                tabindex=\"", "\"\n              ></app-hax-hat-progress>\n            </div>\n        </scrollable-component>\n      </div>\n    "])), this.stepRoutes.map(function (item) {
          return (0, _index.html)(_templateObject5_d77b49e0c0ce11ecbdb54bb05e9423ad || (_templateObject5_d77b49e0c0ce11ecbdb54bb05e9423ad = babelHelpers.taggedTemplateLiteral(["<li\n              ?disabled=\"", "\"\n              class=\"", "\">\n                <a href=\"", "\"\n                  ?disabled=\"", "\"\n                  @click=\"", "\"\n                  id=\"link-", "\"\n                  title=\"", "\"\n                  data-step=\"", "\"\n                  ></a>\n              </li>"])), _this6.step < item.step ? true : false, _this6.step === item.step ? 'active-step' : '', item.path, _this6.step < item.step ? true : false, _this6.stepLinkClick, item.id, item.label, item.step);
        }), this.step !== 1 ? '-1' : '', this.chooseStructure, this.step !== 1 ? '-1' : '', this.chooseStructure, this.renderTypes(this.step), this.appSettings && this.appSettings.themes ? this.themeNames.map(function (themeKey) {
          return (0, _index.html)(_templateObject6_d77b49e0c0ce11ecbdb54bb05e9423ad || (_templateObject6_d77b49e0c0ce11ecbdb54bb05e9423ad = babelHelpers.taggedTemplateLiteral(["\n                <button\n                  aria-label=\"", " theme\"\n                  value=\"", "\"\n                  class=\"theme-button\"\n                  @click=", "\n                  tabindex=\"", "\"\n                >\n                  <img class=\"theme-img\" src=", " alt=\"\" loading=\"lazy\" decoding=\"async\" />\n                  <div>", "</div>\n                </button>\n                "])), _this6.appSettings.themes[themeKey].name, themeKey, _this6.chooseTheme, _this6.step !== 3 ? '-1' : '', _this6.appSettings.themes[themeKey].thumbnail, _this6.appSettings.themes[themeKey].name);
        }) : "", this.step !== 4 ? '-1' : '', this.step !== 4 ? '-1' : '', this.chooseName, this.progressReady, this.progressFinished, this.step !== 5 ? '-1' : '');
      }
    }], [{
      key: "tag",
      get: function get() {
        return 'app-hax-steps';
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread(_objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(AppHaxSteps), "properties", this)), {}, {
          step: {
            type: Number,
            reflect: true
          },
          stepRoutes: {
            type: Array
          },
          themeNames: {
            type: Array
          },
          loaded: {
            type: Boolean,
            reflect: true
          },
          appSettings: {
            type: Object
          }
        });
      }
    }, {
      key: "styles",
      get: function get() {
        return [].concat(babelHelpers.toConsumableArray(babelHelpers.get(babelHelpers.getPrototypeOf(AppHaxSteps), "styles", this)), [(0, _index.css)(_templateObject7_d77b49e0c0ce11ecbdb54bb05e9423ad || (_templateObject7_d77b49e0c0ce11ecbdb54bb05e9423ad = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n          color: var(--app-hax-accent-color, var(--accent-color));\n        }\n        scrollable-component {\n          --scrollbar-width: 0px;\n          --scrollbar-height: 0px;\n          --scrollbar-padding: 0;\n          --viewport-overflow-x: hidden;\n          overflow: hidden;\n        }\n        #grid-container {\n          display: grid;\n          grid-template-columns: 200px 200px;\n          background: transparent;\n        }\n        .carousel-with-snapping-track {\n          display: grid;\n          grid-auto-flow: column;\n          grid-gap: 30px;\n        }\n        .carousel-with-snapping-item {\n          display: flex;\n          flex-direction: column;\n          align-items: center;\n          justify-content: normal;\n          scroll-snap-align: center;\n          scroll-snap-stop: always;\n          width: var(--viewport-width);\n          font-size: 1.5rem;\n          text-align: center;\n          overflow-x: hidden;\n          max-height: 70vh;\n          padding-top: 3vh;\n        }\n        #step-links {\n          padding: 0;\n          margin: 0;\n        }\n        ul,li {\n          list-style: none;\n        }\n        li {\n          display: inline-flex;\n          margin: 4px;\n        }\n        li {\n          border: 1px solid black;\n          border-radius: 50%;\n          background-color: black;\n        }\n        li a {\n          padding: 5px;\n          width: 5px;\n          height: 5px;\n          margin: 0;\n          display: block;\n        }\n        li a[disabled] {\n          pointer-events: none;\n        }\n        li[disabled] {\n          background-color: grey\n        }\n        li.active-step {\n          background-color: orange;\n        }\n        app-hax-button {\n          padding: 10px 0px 10px 0px;\n          background: transparent;\n        }\n        #theme-container {\n          display: flex;\n          flex-direction: row;\n          align-items: center;\n          justify-content: center;\n        }\n        img {\n          pointer-events: none;\n        }\n        #themeContainer {\n          width: 70vw;\n          height: 55vh;\n        }\n        .theme-button {\n          background-color: transparent;\n          border: none;\n          margin: 8px;\n          padding: 8px;\n          width: 245px;\n        }\n        \n        .theme-button div {\n          font-family: 'Press Start 2P', sans-serif;\n          font-size: 16px;\n          margin-top: 16px;\n        }\n        .theme-button:focus,\n        .theme-button:hover {\n          outline: 4px solid var(--app-hax-accent-color, var(--accent-color));\n          outline-offset: 4px;\n          background-color: transparent;\n          border: none;\n          cursor: pointer;\n        }\n        #sitename {\n          font-family: 'Press Start 2P', sans-serif;\n          font-size: 64px;\n          padding: 8px;\n          width: 40vw;\n          text-align: center;\n        }\n        app-hax-site-button {\n          justify-content: center;\n          --app-hax-site-button-width: 30vw;\n          --app-hax-site-button-min-width: 240px;\n        }\n        app-hax-hat-progress {\n          height: 400px;\n          width: 400px;\n          display: block;\n        }\n\n        @media (max-width: 800px) {\n          .theme-button {\n            width: unset;\n            padding: 0;\n          }\n          .theme-button div {\n            font-size: 12px;\n            margin-top: 8px;\n          }\n          .theme-button img {\n            height: 70px;\n          }\n          app-hax-site-button {\n            width: 320px;\n            max-width: 60vw;\n            --app-hax-site-button-font-size: 2.5vw;\n          }\n          #sitename {\n            width: 70vw;\n            font-size: 20px;\n          }\n          #grid-container {\n            grid-template-columns: 150px 150px;\n          }\n        }\n        @media (max-height: 600px) {\n          .carousel-with-snapping-item {\n            padding-top: 4px;\n            max-height: 57vh;\n          }\n          #sitename {\n            width: 40vw;\n            font-size: 14px;\n          }\n          app-hax-hat-progress {\n            transform: scale(.5);\n            margin-top: -18vh;\n          }\n        }\n        @media (max-width: 500px) {\n          app-hax-hat-progress {\n            transform: scale(.5);\n            margin-top: -15vh;\n          }\n        }\n        @media (max-height: 400px) {\n          .carousel-with-snapping-item {\n            padding-top: 4px;\n            max-height: 40vh;\n          }\n          app-hax-hat-progress {\n            transform: scale(.3);\n          }\n          .carousel-with-snapping-item.active-step app-hax-hat-progress {\n            position: fixed;\n            top: 20%;\n            left: 20%;\n          }\n        }\n      "])))]);
      }
    }]);
    return AppHaxSteps;
  }(_simpleColors.SimpleColors);

  _exports.AppHaxSteps = AppHaxSteps;
  customElements.define(AppHaxSteps.tag, AppHaxSteps);
});
