// @lrnwebcomponents/app-hax/lib/v1/app-hax-site-bar.js
import{html as e,css as i}from"../../../../lit/index.js";import"../../../simple-icon/lib/simple-icons.js";import"../../../simple-icon/lib/simple-icon-button-lite.js";import{SimpleColors as t}from"../../../simple-colors/simple-colors.js";import{animate as n}from"../../../../@lit-labs/motion/index.js";export class AppHaxSiteBars extends t{static get tag(){return"app-hax-site-bar"}constructor(){super(),this.icon="add",this.opened=!1,this.inprogress=!1,this.iconLink="https://www.psu.edu"}static get properties(){return{...super.properties,opened:{type:Boolean,reflect:!0},icon:{type:String},inprogress:{type:Boolean,reflect:!0},iconLink:{type:String,attribute:"icon-link"}}}updated(e){super.updated&&super.updated(e),e.forEach(((e,i)=>{"opened"===i&&this.dispatchEvent(new CustomEvent(`${i}-changed`,{detail:{value:this[i]}}))}))}static get styles(){return[...super.styles,i`
        :host {
          --main-banner-width: 500px;
          --main-banner-height: 80px;
          --band-banner-height: 150px;
          display: inline-block;
          background-image: linear-gradient(
            var(--simple-colors-default-theme-accent-9) 80%,
            var(--simple-colors-default-theme-accent-6)
          );
          color: var(--simple-colors-default-theme-accent-1);
        }

        :host([opened]) {
          background-image: linear-gradient(
            var(--simple-colors-default-theme-accent-12),
            var(--simple-colors-default-theme-accent-8)
          );
        }
        #mainCard {
          display: flex;
          flex-direction: row;
          justify-content: space-between;
          align-items: center;
          width: var(--main-banner-width);
          height: var(--main-banner-height);
        }

        #band {
          display: flex;
          flex-direction: column;
          height: 1px;
          width: var(--main-banner-width);
          visibility: none;
          overflow: hidden;
        }

        :host([opened]) #band {
          height: var(--band-banner-height);
          visibility: visible;
        }
        a {
          flex: 1;
        }
        #labels {
          flex: 6;
          overflow: hidden;
          text-overflow: ellipsis;
        }
        #icon {
          --simple-icon-width: 49px;
          --simple-icon-height: 49px;
          color: var(--simple-colors-default-theme-accent-1);
        }
        #dots {
          --simple-icon-width: 49px;
          --simple-icon-height: 49px;
          color: var(--simple-colors-default-theme-accent-1);
        }
      `]}__clickButton(){this.opened=!this.opened}render(){return e`
      <div id="mainCard">
        <a href="${this.iconLink}" tabindex="-1">
          <simple-icon-button-lite icon=${this.icon} id="icon"></simple-icon-button-lite
        ></a>
        <div id="labels">
          <slot name="heading"></slot>
          <slot name="subHeading"></slot>
        </div>
        <simple-icon-button-lite icon="more-vert" id="dots" @click=${this.__clickButton}></simple-icon-button-lite>
      </div>
      <div id="band" ${n()}>
        <slot name="band"></slot>
      </div>
    `}}customElements.define(AppHaxSiteBars.tag,AppHaxSiteBars);
