// @lrnwebcomponents/app-hax/app-hax.js
import{LitElement as e,css as t,html as a}from"../../lit/index.js";import{localStorageSet as o,localStorageGet as s}from"../utils/utils.js";import{toJS as i,autorun as r}from"../../mobx/dist/mobx.esm.js";import{store as n}from"./lib/v1/AppHaxStore.js";import{AppHaxAPI as p}from"./lib/v1/AppHaxBackendAPI.js";import"./lib/v1/AppHaxRouter.js";import"./lib/v1/app-hax-steps.js";import"./lib/v1/app-hax-label.js";import"./lib/v1/app-hax-top-bar.js";import"./lib/v1/app-hax-site-button.js";const l=new URL("./lib/assets/images/HAXLogo.svg",import.meta.url).href;function darkToggle(e){e.matches?n.darkMode=!0:n.darkMode=!1}export class AppHax extends e{static get tag(){return"app-hax"}playSound(e){if(n.soundStatus)try{switch(e){case"click":case"click2":case"coin":case"coin2":case"hit":case"success":this.audio=new Audio(new URL(`./lib/assets/sounds/${e}.mp3`,import.meta.url).href),this.audio.play();break;default:this.audio=new Audio(new URL("./lib/assets/sounds/hit.mp3",import.meta.url).href),this.audio.play(),console.warn(`${e} is not a valid sound file yet`)}}catch(e){console.warn(e)}}connectedCallback(){super.connectedCallback(),window.matchMedia("(prefers-color-scheme: dark)").addEventListener("change",darkToggle)}disconnectedCallback(){window.matchMedia("(prefers-color-scheme: dark)").removeEventListener("change",darkToggle),super.disconnectedCallback()}constructor(){super(),this.userMenuOpen=!1,this.courses=[],this.activeItem={},this.phrases={new:["What's ya name?","HAX to the moon","Welcome to the Jungle","We like to party","Build something awesome","Everything is awesome!","Everything is cool when you're part of the team","When you're living our dream"],return:["Welcome back, take 2?","That wasn't very long","Sup?","You again? Awesome!","Let's do this","There can only be one ring...","There is another","Fancy that, I love HAX and you show up"]},this.isNewUser=null,this.routes=[{path:"createSite-step-1",component:"fake",step:1,id:"step-1",label:"Welcome",statement:"Let's start a new journey!"},{path:"createSite-step-2",component:"fake",step:2,id:"step-2",label:"Pick type",statement:"What type of journey is this?"},{path:"createSite-step-3",component:"fake",step:3,id:"step-3",label:"Theme Select",statement:"What's it look like?"},{path:"createSite-step-4",component:"fake",step:4,id:"step-4",label:"Name",statement:"Name this journey"},{path:"createSite-step-5",component:"fake",step:5,id:"step-5",label:"Loading",statement:"Get ready for your journey"},{path:"/",component:"fake",name:"home",label:"Welcome back",statement:"Let's explore HAX land"},{path:"/search",component:"fake-search-e",name:"search",label:"Search",statement:"Discover active adventures"},{path:"/(.*)",component:"fake",name:"404",label:"404 :[",statement:"it's not you.. it's me"}],this.searchTerm="",this.appMode="",this.soundIcon="",this.store=n,this.source="",this.sound=new Audio,this.source=new URL("../demo/sites.json",import.meta.url).href,import("../simple-modal/simple-modal.js"),r((()=>{this.isNewUser=i(n.isNewUser),this.isNewUser&&"create"!==this.appMode&&(this.appMode="create",setTimeout((()=>{n.createSiteSteps=!0}),0))})),r((()=>{this.userName=i(n.user.name)})),r((()=>{this.searchTerm=i(n.searchTerm)})),r((async()=>{if(n.location&&n.location.route){const e=i(n.location);e.route.step?(this.appMode="create",setTimeout((()=>{n.createSiteSteps=!0}),0)):"404"===e.route.name?(n.createSiteSteps=!1,this.appMode="404",setTimeout((()=>{n.toast("the page miss.. it burns!!!",3e3,{fire:!0,walking:!0})}),500)):"home"===e.route.name||"search"===e.route.name?this.appMode="home":console.warn(e.route)}})),r((()=>{n.routes.length>0&&null===n.location&&(n.location=i(n.routes[0]))})),null===s("app-hax-darkMode",null)&&(n.darkMode=window.matchMedia("(prefers-color-scheme: dark)").matches),r((()=>{o("app-hax-darkMode",i(n.darkMode)),i(n.darkMode)?(document.body.classList.add("dark-mode"),n.toast("I'm ascared of the dark",2e3,{fire:!0})):(document.body.classList.remove("dark-mode"),n.toast("Sunny day it is",2e3,{hat:"random"}))})),r((()=>{const e=i(n.appMode);e&&(document.body.classList.remove("app-hax-search"),document.body.classList.remove("app-hax-create"),document.body.classList.remove("app-hax-404"),document.body.classList.remove("app-hax-home"),document.body.classList.add(`app-hax-${e}`))})),r((()=>{if(i(n.appReady)&&i(n.isLoggedIn)){console.log("I am ready to get sites list");i(n.appSettings);setTimeout((async()=>{const e=await p.makeCall("getSitesList");n.manifest=e,console.log(`Manifest Length: ${n.manifest.items.length}`)}),0)}else i(n.appReady)&&!i(n.isLoggedIn)&&setTimeout((()=>{console.log("We in App-Hax - Running Log in"),this.login()}),0)}))}static get properties(){return{courses:{type:Array},source:{type:String},userName:{type:String},activeItem:{type:Object},soundIcon:{type:String},routes:{type:Array},searchTerm:{type:String},appMode:{type:String},isNewUser:{type:Boolean},phrases:{type:Object},userMenuOpen:{type:Boolean}}}getData(){fetch(this.source).then((e=>e.json())).then((e=>{this.courses=[],e.items.forEach((e=>{const t={title:e.title,description:e.description,link:i(n.sitesBase).concat(e.slug),icon:e.metadata.theme.variables?e.metadata.theme.variables.icon:""};this.courses.push(t)}))}))}reset(e=!0){try{window.localStorage.removeItem("app-hax-step"),window.localStorage.removeItem("app-hax-darkMode"),window.localStorage.removeItem("app-hax-soundStatus"),window.localStorage.removeItem("app-hax-site"),e&&(window.location=document.querySelector("base").href)}catch(e){console.warn(e)}}login(){console.log("Login Function in App-HAX ran"),import("./lib/v1/app-hax-site-login.js").then((()=>{const e=document.createElement("app-hax-site-login");if(this.querySelector('[slot="externalproviders"]')){const t=this.querySelector('[slot="externalproviders"]').cloneNode(!0);e.appendChild(t)}const t=new CustomEvent("simple-modal-show",{bubbles:!0,cancelable:!0,detail:{title:"Character select",modal:!0,elements:{content:e},modal:!0,invokedBy:this,styles:{"--simple-modal-titlebar-background":"orange","--simple-modal-titlebar-color":"black","--simple-modal-width":"25vw","--simple-modal-min-width":"300px","--simple-modal-z-index":"100000000","--simple-modal-height":"40vh","--simple-modal-min-height":"400px","--simple-modal-titlebar-height":"40px"}}});this.dispatchEvent(t)}))}static get styles(){return[t`
        :host {
          display: block;
        }
        .topbar-character {
          transform: scale(0.4, 0.4);
          margin: -36px -35px 0px 0px;
          vertical-align: text-top;
          position: fixed;
          top: 0px;
          right: 0px;
          overflow: hidden;
          height: 120px;
          cursor: pointer;
        }
        .content {
          text-align: center;
          margin-top: 32px;
        }
        .four04-character {
          margin-top: 100px;
          transform: scale(2);
        }
        .start-journey {
          display: flex;
          justify-content: center;
        }
        app-hax-top-bar {
          top: 0;
          z-index: 1000;
          right: 0;
          left: 0;
          position: fixed;
        }
        .label {
          text-align: center;
          margin-top: 48px;
        }
        app-hax-label {
          animation: 0.8s ease-in-out 0s scrollin;
          -webkit-animation: 0.8s ease-in-out 0s scrollin;
          display: block;
        }
        @media (prefers-reduced-motion: reduce) {
          app-hax-label {
            animation: none;
            -webkit-animation: none;
          }
        }
        app-hax-label h1 {
          font-weight: normal;
          font-size: 4vw;
          margin: 0;
          padding: 0;
        }
        @keyframes scrollin {
          from {
            margin-top: -240px;
            margin-bottom: 240px;
          }
          to {
            margin-top: 0;
            margin-bottom: 0;
          }
        }
        .haxLogo {
          height: 40px;
          width: 40px;
          background-color: white;
          margin: 4px;
        }
        .space-hack {
          display: inline-flex;
          width: 64px;
          height: 48px;
        }
        .soundToggle {
          margin-right: 16px;
          position: relative;
          display: inline-flex;
          vertical-align: top;
        }
        .soundToggle img {
          width: 24px;
          height: 24px;
        }
        app-hax-search-bar {
          vertical-align: middle;
          display: inline-flex;
        }
        main {
          margin-top: 64px;
        }
        .user-menu {
          display: none;
        }
        .user-menu.open {
          display: block;
          top: 48px;
          right: 0px;
          position: absolute;
          border: 1px solid black;
          background-color: white;
        }
        .user-menu.open button {
          display: block;
          width: 100%;
          margin: 0;
          padding: 8px;
          font-size: 20px;
          font-family: "Press Start 2P", sans-serif;
          color: black;
          background-color: white;
        }
        .user-menu.open button:hover,
        .user-menu.open button:active,
        .user-menu.open button:focus {
          background-color: black;
          color: white;
        }
        random-word {
          transform: rotate(25deg);
          position: absolute;
          right: 10px;
          top: 120px;
          padding: 12px;
          font-size: 12px;
          border: 4px solid black;
          background-color: yellow;
          color: black;
          width: 100px;
          word-wrap: break-word;
          text-align: center;
          cursor: pointer;
          user-select: none;
          opacity: 1;
          visibility: visible;
          transition: all 0.3s ease-in-out;
        }
        @media (max-width: 680px) {
          random-word {
            visibility: none;
            opacity: 0;
          }
        }
      `]}updated(e){super.updated&&super.updated(e),e.forEach(((e,t)=>{["routes","appMode"].includes(t)&&(n[t]=this[t])}))}firstUpdated(e){super.firstUpdated&&super.firstUpdated(e),import("../../wired-elements/lib/wired-button.js"),import("./lib/v1/app-hax-toast.js"),import("./lib/v1/app-hax-wired-toggle.js"),import("./lib/v1/app-hax-search-bar.js"),import("./lib/v1/app-hax-search-results.js"),import("../rpg-character/rpg-character.js"),import("./lib/random-word/random-word.js"),this.dispatchEvent(new CustomEvent("app-hax-loaded",{composed:!0,bubbles:!0,cancelable:!1,detail:!0})),n.appReady=!0,r((()=>{i(n.appReady)?document.body.classList.add("app-loaded"):document.body.classList.remove("app-loaded")})),this.shadowRoot.querySelector("app-hax-label").addEventListener("animationend",(e=>{"scrollin"===e.animationName&&n.appEl.playSound("coin2")})),n.appEl=this,r((()=>{n.activeItem&&(this.activeItem=i(n.activeItem))})),r((()=>{this.soundIcon=i(n.soundStatus)?new URL("./lib/assets/images/FullVolume.svg",import.meta.url).href:new URL("./lib/assets/images/Silence.svg",import.meta.url).href,i(n.soundStatus)?n.toast("Can you hear me now? Good.",2e3,{hat:"random"}):n.toast("Sound off.. hey.. HELLO!?!",2e3,{fire:!0})}))}soundToggle(e){n.soundStatus=!i(n.soundStatus),o("app-hax-soundStatus",i(n.soundStatus))}toggleMenu(){this.userMenuOpen=!this.userMenuOpen}render(){return a`<app-hax-router></app-hax-router>
      <header>
        <app-hax-top-bar>
          <img
            class="haxLogo"
            src="${l}"
            slot="left"
            alt=""
            loading="lazy"
            decoding="async"
          />
          <app-hax-search-bar
            slot="center"
            ?disabled="${this.isNewUser}"
          ></app-hax-search-bar>
          <wired-button
            elevation="1"
            slot="right"
            class="soundToggle"
            @click="${this.soundToggle}"
          >
            <img
              src="${this.soundIcon}"
              alt=""
              loading="lazy"
              decoding="async"
            />
          </wired-button>
          <app-hax-wired-toggle slot="right"></app-hax-wired-toggle>
          <div class="space-hack" slot="right"></div>
          <rpg-character
            class="topbar-character"
            seed="${this.userName}"
            slot="right"
            @click="${this.toggleMenu}"
          ></rpg-character>
          <div
            slot="right"
            class="user-menu ${this.userMenuOpen?"open":""}"
          >
            <button>Site settings</button>
            <button>Site outline</button>
            <button>New Journey</button>
            <button>Account info</button>
            <button>log out</button>
          </div>
        </app-hax-top-bar>
      </header>
      <main>
        <div class="label">
          <app-hax-label>
            ${this.activeItem?a`
                  <h1>${this.activeItem.label}</h1>
                  <div slot="subtitle">${this.activeItem.statement}</div>
                `:""}
          </app-hax-label>
        </div>
        <random-word
          key="${this.isNewUser?"new":"return"}"
          .phrases="${this.phrases}"
          @click="${this.getNewWord}"
        ></random-word>
        <section class="content">${this.appBody(this.appMode)}</section>
      </main>`}getNewWord(){this.shadowRoot.querySelector("random-word").getNewWord()}appBody(e){switch(e){case"home":case"search":return this.templateHome();case"create":return this.templateCreate();default:return this.template404()}}templateHome(){return a` ${this.searchTerm?"":a` <div class="start-journey">
            <a
              href="createSite-step-1"
              @click="${this.startJourney}"
              tabindex="-1"
            >
              <app-hax-site-button
                label="> Start new journey"
              ></app-hax-site-button>
            </a>
          </div>`}

      <app-hax-search-results></app-hax-search-results>`}templateCreate(){return a`<app-hax-steps></app-hax-steps>`}template404(){return a` <div class="four04">
      <a
        href="createSite-step-1"
        class="start-journey"
        @click="${this.startJourney}"
        tabindex="-1"
      >
        <app-hax-site-button
          label="> Start a new journey"
        ></app-hax-site-button>
      </a>
      <rpg-character
        class="four04-character"
        fire
        walking
        seed="${this.userName}"
      ></rpg-character>
    </div>`}startJourney(e){n.step=1,this.appMode="create"}}customElements.define(AppHax.tag,AppHax),window.AppHax=window.AppHax||{},window.AppHax.requestAvailability=()=>(window.AppHax.instance||(window.AppHax.instance=document.querySelector(AppHax.tag)),window.AppHax.instance);
